pyro/optim/dct_adam.py

```diff
@@ -73,44 +94,98 @@ def step(self, closure=None):
 
-                state['step'] += 1
+        state = self.state[p]
 
-                # Decay the first and second moment running average coefficient
-                exp_avg.mul_(beta1).add_(1 - beta1, grad)
-                exp_avg_sq.mul_(beta2).addcmul_(1 - beta2, grad, grad)
+        # State initialization
+        if len(state) == 0:
+            state['step'] = 0
+            # Exponential moving average of gradient values
+            state['exp_avg'] = torch.zeros_like(grad)
+            # Exponential moving average of squared gradient values
+            state['exp_avg_sq'] = torch.zeros_like(grad)
 
-                denom = exp_avg_sq.sqrt().add_(group['eps'])
+        exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
+        beta1, beta2 = group['betas']
 
-                bias_correction1 = 1 - beta1 ** state['step']
-                bias_correction2 = 1 - beta2 ** state['step']
-                step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1
+        state['step'] += 1
 
-                if time_dim is None:
-                    p.data.addcdiv_(-step_size, exp_avg, denom)
-                else:
-                    step = _transform_inverse(exp_avg / denom, time_dim, duration)
-                    p.data.add_(step.mul_(-step_size))
+        # Decay the first and second moment running average coefficient
+        exp_avg.mul_(beta1).add_(1 - beta1, grad)
+        exp_avg_sq.mul_(beta2).addcmul_(1 - beta2, grad, grad)
 
-        return loss
+        denom = exp_avg_sq.sqrt().add_(group['eps'])
+
+        bias_correction1 = 1 - beta1 ** state['step']
+        bias_correction2 = 1 - beta2 ** state['step']
+        step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1
+
+        if time_dim is None:
+            p.data.addcdiv_(-step_size, exp_avg, denom)
+        else:
+            step = _transform_inverse(exp_avg / denom, time_dim, duration)
+            p.data.add_(step.mul_(-step_size))
+
+    def _step_param_subsample(self, group, p, subsample):
+        mask = _get_mask(p, subsample)
+
+        grad = p.grad.data.masked_select(mask)
+        grad.clamp_(-group['clip_norm'], group['clip_norm'])
+
+        # Transform selected parameters via dct.
+        time_dim = getattr(p, "_pyro_dct_dim", None)
+        if time_dim is not None:
+            duration = p.size(time_dim)
+            grad = _transform_forward(grad, time_dim, duration)
+
+        state = self.state[p]
+
+        # State initialization
+        if len(state) == 0:
+            state['step'] = torch.zeros_like(p)
+            # Exponential moving average of gradient values
+            state['exp_avg'] = torch.zeros_like(p)
+            # Exponential moving average of squared gradient values
+            state['exp_avg_sq'] = torch.zeros_like(p)
+
+        beta1, beta2 = group['betas']
+
+        state_step = state['step'].masked_select(mask).add_(1)
+        state['step'].masked_scatter_(mask, state_step)
```

Comment:
Interesting pattern! Consider abstracting this out as a `masked_add_()` helper in `pyro.distributions.util`.

Reply:
I would like to skip this abstraction. At other places, we not only use `add` but also `mul` or `addcmul`.